modules/stochastic_tools/include/surrogates/LibtorchSimpleNNTrainer.h

```diff
@@ -0,0 +1,98 @@
+//* This file is part of the MOOSE framework
+//* https://www.mooseframework.org
+//*
+//* All rights reserved, see COPYRIGHT for full restrictions
+//* https://github.com/idaholab/moose/blob/master/COPYRIGHT
+//*
+//* Licensed under LGPL 2.1, please see LICENSE for details
+//* https://www.gnu.org/licenses/lgpl-2.1.html
+
+#pragma once
+
+#ifdef TORCH_ENABLED
+#include <torch/torch.h>
+#include "LibtorchSimpleNeuralNet.h"
+#endif
+
+#include "libmesh/utility.h"
+#include "SurrogateTrainer.h"
+
+class LibtorchSimpleNNTrainer : public SurrogateTrainer
+{
+public:
+  static InputParameters validParams();
+
+  LibtorchSimpleNNTrainer(const InputParameters & parameters);
+
+  virtual void preTrain() override;
+
+  virtual void train() override;
+
+  virtual void postTrain() override;
+
+protected:
+#ifdef TORCH_ENABLED
+
+  // A custom strcuture which is used to organize data foor the training of
+  // torch-based neural nets.
+  struct MyData : torch::data::datasets::Dataset<MyData>
+  {
+  public:
+    MyData(torch::Tensor dt, torch::Tensor rt) : _data_tensor(dt), _response_tensor(rt) {}
+    torch::data::Example<> get(size_t index) override
+    {
+      return {_data_tensor.slice(0, index, index + 1), _response_tensor[index]};
+    }
+
+    torch::optional<size_t> size() const override { return _response_tensor.sizes()[0]; }
+
+  private:
+    torch::Tensor _data_tensor;
+    torch::Tensor _response_tensor;
+  };
+
+#endif
+
+private:
+  /// Data from the current sampler row
+  const std::vector<Real> & _sampler_row;
+
+  /// Response value
+  const Real & _response;
+
+  /// The gathered data in a flattened form to be able to convert easily to torch::Tensor.
+  std::vector<Real> _flattened_data;
+
+  /// The gathered response in a flattened form to be able to convert easily to torch::Tensor.
+  std::vector<Real> _flattened_response;
+
+  /// Number of batches we want to prepare
+  unsigned int _no_batches;
+
+  /// Number of epochs for the training
+  unsigned int _no_epocs;
```

Comment:
Does pytorch allow us to specify something like "relative tol " instead of the number of iterations, `_no_epocs`?

Reply:
I added the option in the training loop!